### PR TITLE
Alexseev Danila. 1 lab option №3

### PR DESCRIPTION
**Pull request by lxvdnl**
I made a plugin that throws an error if the function name contains 'deprecated'

### CODE REVIEW

File: clang/labs/lab1/alexseev_danila/deprecatedWarning.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+class DepFuncVisitor : public RecursiveASTVisitor<DepFuncVisitor> {
+
+private:
+
+    ASTContext *Context;
+
+public:
+
+    explicit DepFuncVisitor(ASTContext *context) : Context(context) {}
```
Comment (m-ly4) [inline clang/labs/lab1/alexseev_danila/deprecatedWarning.cpp]: Please fix diffs in clang-format check and warnings in clang-tidy check.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by lxvdnl** [comment]
I've fixed clang-format and clang-tidy

**Review by aobolensk — APPROVED** [review]

File: clang/labs/lab1/alexseev_danila/deprecatedWarning.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+class DepFuncConsumer : public ASTConsumer {
+private:
+  CompilerInstance &Instance;
+
+public:
+  explicit DepFuncConsumer(CompilerInstance &CI) : Instance(CI) {}
+
+  void HandleTranslationUnit(ASTContext &Context) override {
+    DepFuncVisitor Visitor(&Instance.getASTContext());
```
Comment (m-ly4) [inline clang/labs/lab1/alexseev_danila/deprecatedWarning.cpp]: Context from input argument is enough. No need to store Instance for that.

File: clang/labs/lab1/alexseev_danila/deprecatedWarning.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+protected:
+  std::unique_ptr<ASTConsumer>
+  CreateASTConsumer(CompilerInstance &Compiler,
+                    llvm::StringRef InFile) override {
+    return std::make_unique<DepFuncConsumer>(Compiler);
+  }
+
+  bool ParseArgs(const CompilerInstance &Compiler,
+                 const std::vector<std::string> &Args) override {
+    return true;
```
Comment (m-ly4) [inline clang/labs/lab1/alexseev_danila/deprecatedWarning.cpp:61]: Please add an option specifying a function name to exclude from check.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/labs/lab1/alexseev_danila/deprecatedWarning.cpp (hunk 3 above)
Comment (lxvdnl) [reply clang/labs/lab1/alexseev_danila/deprecatedWarning.cpp:61]: I added the '-exclude' option which takes an argument (function name). Example: '-exclude=deprecatedFunc'. When this option is used, a function named 'deprecatedFunc' will not generate a warning. I also added a check of the option’s operation to the test.

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**